tests/test_progress.py

```diff
@@ -0,0 +1,210 @@
+        """
+        result = self.runner.invoke(
+            self.get_cmd(logging.INFO, enabled=True),
+            ["--progress"],
+        )
+        self.assertEqual(result.exit_code, 0, clickResultMsg(result))
+
+    def test_click_disabled_globally(self):
+        """Test turning on progress in click commands.
+        """
+        result = self.runner.invoke(
+            self.get_cmd(logging.INFO, enabled=False),
+            ["--no-progress"],
+        )
+        self.assertEqual(result.exit_code, 0, clickResultMsg(result))
+
+    def test_click_disabled_by_log_level(self):
+        """Test that progress reports below the current log level are disabled,
+        even if progress is globally enabled.
+        """
+        result = self.runner.invoke(
+            self.get_cmd(logging.DEBUG, enabled=False),
+            ["--progress"],
+        )
+        self.assertEqual(result.exit_code, 0, clickResultMsg(result))
+
+
+class MockedProgressHandlerTestCase(unittest.TestCase):
+    """Test that the interface layer for progress reporting works by using
+    mock handler and progress bar objects.
+    """
+
+    def setUp(self):
+        self.logger = logging.getLogger("test_progress")
+        self.logger.setLevel(logging.INFO)
+        Progress.set_handler(MockProgressHandler())
+        self.progress = Progress("test_progress")
+
+    def tearDown(self):
+        MockProgressHandler.last = None
+        Progress.set_handler(None)
+        self.logger.setLevel(logging.NOTSET)
+
+    def test_bar_iterable(self):
+        """Test using `Progress.bar` to wrap an iterable.
+        """
+        iterable = list(range(5))
+        with self.progress.bar(iterable) as bar:
+            r = list(bar)
+        self.assertEqual(r, iterable)
+        self.assertEqual(iterable + [len(iterable)], bar.reported)
+
+    def test_bar_update(self):
+        """Test using `Progress.bar` with manual updates.
+        """
+        with self.progress.bar(total=10) as bar:
+            for i in range(5):
+                bar.update(2)
+        self.assertEqual(list(range(0, 12, 2)), bar.reported)
+
+    def test_iter_chunks(self):
+        """Test using `Progress.iter_chunks`.
+        """
+        iterable = [list(range(2)), list(range(3))]
+        seen = []
+        for chunk in self.progress.iter_chunks(iterable):
+            seen.extend(chunk)
+        self.assertEqual(seen, iterable[0] + iterable[1])
+        self.assertEqual(MockProgressBar.last.reported, [0, 2, 5])
+
+    def test_iter_item_chunks(self):
+        """Test using `Progress.iter_item_chunks`.
+        """
+        mapping = {"x": list(range(2)), "y": list(range(3))}
+        seen = {}
+        for key, chunk in self.progress.iter_item_chunks(mapping.items()):
+            seen[key] = chunk
+        self.assertEqual(seen, mapping)
+        self.assertEqual(MockProgressBar.last.reported, [0, 2, 5])
+
+
+if __name__ == "__main__":
+    unittest.main()
+    print(Progress._setters)
```

Comment:
Maybe say this is for debugging? pytest won't ever see it.

Reply:
I had actually intended to remove it.  Will do that.